contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,309 @@
+            if (totalRewards > lastAvailableRewards) {
+                // Traverse all the eligible services and adjust their rewards proportional to leftovers
+                totalRewards = 0;
+                for (uint256 i = 0; i < numServices; ++i) {
+                    // Calculate the updated reward
+                    uint256 updatedReward = (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
+                    // Add to the total updated reward
+                    totalRewards += updatedReward;
+                    // Add reward to the service overall reward
+                    uint256 curServiceId = eligibleServiceIds[i];
+                    mapServiceInfo[curServiceId].reward += updatedReward;
+                }
+
+                // If the reward adjustment happened to have small leftovers, add it to the last traversed service
+                if (lastAvailableRewards > totalRewards) {
+                    mapServiceInfo[numServices - 1].reward += lastAvailableRewards - totalRewards;
+                }
+                // Set available rewards to zero
+                lastAvailableRewards = 0;
+            } else {
+                // Traverse all the eligible services and add to their rewards
+                for (uint256 i = 0; i < numServices; ++i) {
+                    // Add reward to the service overall reward
+                    uint256 curServiceId = eligibleServiceIds[i];
+                    mapServiceInfo[curServiceId].reward += eligibleServiceRewards[i];
+                }
+
+                // Adjust available rewards
+                lastAvailableRewards -= totalRewards;
+            }
+
+            // Update the storage value of available rewards
+            availableRewards = lastAvailableRewards;
+
+            // Updated current service nonces
+            for (uint256 i = 0; i < size; ++i) {
+                // Get the current service Id
+                uint256 curServiceId = serviceIds[i];
+                mapServiceInfo[curServiceId].nonce = serviceNonces[i];
+            }
+        }
+
+        // Record the current timestamp such that next calculations start from this point of time
+        tsCheckpoint = block.timestamp;
+
+        emit Checkpoint(lastAvailableRewards, numServices);
+    }
+
+    /// @dev Public checkpoint function to allocate rewards up until a current time.
+    function checkpoint() external {
+        _checkpoint(0);
+    }
+
+    /// @dev Unstakes the service.
+    /// @param serviceId Service Id.
+    function unstake(uint256 serviceId) external {
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        // Check for the service ownership
+        if (msg.sender != sInfo.owner) {
+            revert OwnerOnly(msg.sender, sInfo.owner);
+        }
+
+        // Call the checkpoint and get the service index in the set of services
+        uint256 idx = _checkpoint(serviceId);
```

Comment:
I think there should be a simple call to `_checkpoint()` and any service specific stuff to serviceID should happen outside this context. That keeps the code simpler to reason about.

Reply:
As stated above, if there's a way not to spend more gas on figuring out the service Id index, I'd do it.

Reply:
It's several times per day the trade off between `pretty` and `gas efficient`.

Reply:
ok I've refactored it, but still not super happy, will see in the morning